### PR TITLE
(WIP) Add maxrequestbody directive

### PR DESCRIPTION
**Pull request by Ngo-The-Trung**
@mholt I'm looking for opinions on this implementation before I proceed further

[Issue](https://github.com/mholt/caddy/issues/1027)

### CODE REVIEW

File: caddyhttp/httpserver/siteconfig.go [hunk 1]
```diff
@@ -30,6 +30,14 @@ type SiteConfig struct {
 	// standardized way of loading files from disk
 	// for a request.
 	HiddenFiles []string
+
+	// Max amount of bytes a request can send on a given path
```
Comment (mholt) [inline caddyhttp/httpserver/siteconfig.go]: "Max body size for requests to their respective paths"

File: caddyhttp/httpserver/siteconfig.go [hunk 2]
```diff
@@ -30,6 +30,14 @@ type SiteConfig struct {
...
+	MaxRequestBodySizes []PathLimit
+}
+
+type PathLimit struct {
```
Comment (mholt) [inline caddyhttp/httpserver/siteconfig.go]: This'll need some godoc of course

File: caddyhttp/httpserver/server.go [hunk 3]
```diff
@@ -406,3 +417,14 @@ func WriteTextResponse(w http.ResponseWriter, status int, body string) {
 	w.WriteHeader(status)
 	w.Write([]byte(body))
 }
+
+// HandleCommonError checks if the given error is an error that could pop up
+// when handling request. Currently only checks for MaxBytesReader error.
+// Returns a boolean - whether the given error is a common error, and a int -
+// the suggested HTTP status code to return to the user
+func IsCommonError(err error) (bool, int) {
+	if err.Error() == "http: request body too large" {
```
Comment (mholt) [inline caddyhttp/httpserver/server.go]: Reading the string content of an error value is a little scary -- I like to avoid this where possible. It could change any time.

After looking into MaxBytesReader, I see it [makes use of an unexported error value](https://github.com/golang/go/blob/89c4cbd7acc36cf627746f379f0cc002b5b60383/src/net/http/request.go#L982), hence the need for this function which has to read the error string...

Finding https://github.com/golang/go/issues/8479, it seems like the Go team isn't interested in having the error value exported and reused, but the implementation of MaxBytesReader is trivial; we could just borrow from it and change it to meet our needs by using a pre-defined error value. Also it would be useful to mention how big the read body was, just so it's easier to read in the logs.

However, we should decide if we want to centralize the use of a MaxBodyReader like this, or if each middleware that reads the response body should use it instead. Either way, they'll have to check for the kind of error value returned and, if an EntityTooLargeError, they return a special status code in that case.

File: caddyhttp/maxrequestbody/maxrequestbody.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+import (
+	"regexp"
+	"strconv"
+
+	"github.com/mholt/caddy"
+	"github.com/mholt/caddy/caddyhttp/httpserver"
+)
+
+var SIZE_RE = regexp.MustCompile("[1-9][0-9]*[kmKM]?")
```
Comment (mholt) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: To prevent maintenance headaches in the future, can we not use a regular expression? :) This is a simple job for parsing. If it's a positive integer, use it as bytes. If it's a string suffixed with any case variation of KB, MB, (or GB, I suppose), split it out and parse the non-negative number before it.

File: caddyhttp/maxrequestbody/maxrequestbody.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+		mult = 1024
+		arg = arg[0:L]
+	} else if arg[L] == 'M' {
+		mult = 1024 * 1024
+		arg = arg[0:L]
+	}
+	if size, err := strconv.ParseInt(arg, 10, 64); err != nil {
+		return c.ArgErr()
+	} else {
+		config.MaxRequestBodySizes = append(config.MaxRequestBodySizes, httpserver.PathLimit{path, size * mult})
```
Comment (tw4452852) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: Should we check the possible path duplication?

File: caddyhttp/httpserver/server.go [hunk 6]
```diff
@@ -265,6 +265,17 @@ func (s *Server) serveHTTP(w http.ResponseWriter, r *http.Request) (int, error)
 		}
 	}
 
+	// Apply the path-based request body size limit
+	// The error returned by MaxBytesReader is meant to be handled
+	// by whichever middleware/plugin that receives it when calling
+	// .Read() or a similar method on the request body
+	for _, pathlimit := range vhost.MaxRequestBodySizes {
+		if Path(r.URL.Path).Matches(pathlimit.Path) {
+			r.Body = http.MaxBytesReader(w, r.Body, pathlimit.Limit)
```
Comment (tw4452852) [inline caddyhttp/httpserver/server.go]: `r.Body` may be `nil`.

File: caddyhttp/httpserver/server.go (hunk 3 above)
Comment (tw4452852) [inline caddyhttp/httpserver/server.go]: Agree with @mholt , Use the type assertion instead.

File: caddyhttp/httpserver/server.go (hunk 6 above)
Comment (mholt) [inline caddyhttp/httpserver/server.go]: I've heard this debated before, and so I'm not sure if that's true -- but I definitely would not oppose to a nil check here!

File: caddyhttp/maxrequestbody/maxrequestbody.go (hunk 5 above)
Comment (Ngo-The-Trung) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: I've been meaning to ask this - do we:
- clean up the path (add slashes)
- sort the paths (so longer/more specific paths are checked first)

File: caddyhttp/maxrequestbody/maxrequestbody.go (hunk 4 above)
Comment (Ngo-The-Trung) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: Okay

File: caddyhttp/maxrequestbody/maxrequestbody.go (hunk 5 above)
Comment (tw4452852) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: I think we could clean up the path here and if the path exists, overwrite.
Also sorting paths for specific path matching above makes sense.

Comment (mholt) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: Hmm, I thought there was a facility in the `httpserver` package to help with the longest-path-matching thing, but maybe it was only the `proxy` middleware that got that blessed feature. (Man, I still really think we need to totally standardize the code to match requests. But that's a different topic.) Check out the proxy middleware maybe...

File: caddyhttp/httpserver/plugin.go [hunk 7]
```diff
@@ -437,6 +437,7 @@ var directives = []string{
 	"hugo",        // github.com/hacdias/caddy-hugo
 	"mailout",     // github.com/SchumacherFM/mailout
 	"awslambda",   // github.com/coopernurse/caddy-awslambda
+	"maxrequestbody",
```
Comment (mholt) [inline caddyhttp/httpserver/plugin.go]: Also I wonder if it should be called `max_req_body` or something a little shorter (and readable).

Comment (Ngo-The-Trung) [inline caddyhttp/httpserver/plugin.go]: Doesn't seem like anything other directives use underscore at the moment.

Also I will leave renaming last.

**Comment by Ngo-The-Trung** [comment]
Updated. Addressed the comments above (except for the directive name, that comes later)

- Copied MaxBytesReader from the Go std here and use type checking for error
- Updated parsing to allow multiple paths (such as `maxrequestbody / 100MB /index 200GB`)
- Allow sizes of B, KB, MB, GB (optional B, case insensitive) 
- Parse sorts by path lengths
- Add some parsing tests

I don't understand the CI failures.

File: caddyhttp/maxrequestbody/maxrequestbody_test.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	for caseNum, c := range cases {
+		output, err := parseArguments(c.arguments)
+		if c.hasError && (err == nil) {
+			t.Errorf("Expecting error for case %v but none encountered", caseNum)
+		}
+		if !c.hasError && (err != nil) {
+			t.Errorf("Expecting no error for case %v but encountered %v", caseNum, err)
+		}
+
+		if len(c.expected) != len(output) {
```
Comment (tw4452852) [inline caddyhttp/maxrequestbody/maxrequestbody_test.go]: Maybe `reflect.DeepEqual` helps.

File: caddyhttp/maxrequestbody/maxrequestbody_test.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+		// Parse errors
+		{[]PathLimit{}, []PathLimit{}},
+		{[]PathLimit{{"/index", 100}}, []PathLimit{{"/index", 100}}},
+		{[]PathLimit{{"/static", 1}, {"/static/images", 100}, {"/index", 200}}, []PathLimit{{"/static/images", 100}, {"/static", 1}, {"/index", 200}}},
+	}
+
+	for caseNum, c := range cases {
+		output := append([]PathLimit{}, c.arguments...)
+		SortPathLimits(output)
+		if len(c.expected) != len(output) {
```
Comment (tw4452852) [inline caddyhttp/maxrequestbody/maxrequestbody_test.go]: Ditto.

File: caddyhttp/maxrequestbody/maxrequestbody_test.go [hunk 10]
```diff
@@ -0,0 +1,83 @@
+package maxrequestbody
+
+import (
+	"testing"
+
+	. "github.com/mholt/caddy/caddyhttp/httpserver"
```
Comment (tw4452852) [inline caddyhttp/maxrequestbody/maxrequestbody_test.go]: Explicit import may be better.

File: caddyhttp/maxrequestbody/maxrequestbody_test.go [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+)
+
+const (
```
Comment (tw4452852) [inline caddyhttp/maxrequestbody/maxrequestbody_test.go]: Could these be combined with `validUnits` in `maxrequestbody.go` ?

File: caddyhttp/maxrequestbody/maxrequestbody.go [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+	args := c.RemainingArgs()
+	pathLimit, err := parseArguments(args)
+	if err != nil {
+		return c.ArgErr()
+	}
+
+	SortPathLimits(pathLimit)
+
+	config.MaxRequestBodySizes = pathLimit
+	fmt.Println(pathLimit)
```
Comment (tw4452852) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: Debug code?

File: caddyhttp/maxrequestbody/maxrequestbody.go [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+
+	return nil
+}
+
+func parseArguments(args []string) ([]httpserver.PathLimit, error) {
+	pathLimit := []httpserver.PathLimit{}
+	// Arguments appear in pairs of "<path> <limit>"
+	// If there's an argument that's not paired, it's assumed to be
+	// the limit used for all remaining paths ("/")
+	if len(args)%2 == 1 {
```
Comment (tw4452852) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: This could be wrong. Suppose we have this sequence: `/a 100KB 200KB`  or `/a 100KB /b`

File: caddyhttp/httpserver/server.go [hunk 14]
```diff
@@ -406,3 +488,14 @@ func WriteTextResponse(w http.ResponseWriter, status int, body string) {
...
+
+// HandleCommonError checks if the given error is an error that could pop up
+// when handling request. Currently only checks for MaxBytesReader error.
+// Returns a boolean - whether the given error is a common error, and a int -
+// the suggested HTTP status code to return to the user
+func IsCommonError(err error) (bool, int) {
+	if _, ok := err.(MaxBytesExceeded); ok {
+		return true, http.StatusRequestEntityTooLarge
+	}
+	return false, -1
```
Comment (tw4452852) [inline caddyhttp/httpserver/server.go]: `-1` maybe a invalid HTTP status.

File: caddyhttp/httpserver/server.go [hunk 15]
```diff
@@ -392,6 +406,74 @@ func (ln tcpKeepAliveListener) File() (*os.File, error) {
 	return ln.TCPListener.File()
 }
 
+// MaxBytesExceeded is the error type returned by MaxBytesReader
+// when the request body exceeds the limit imposed
+type MaxBytesExceeded int
```
Comment (tw4452852) [inline caddyhttp/httpserver/server.go]: Is there any reason for using a `int` here ?

**Comment by tw4452852** [comment]
@Ngo-The-Trung 
Also for CI issue, you should check `TestStandardPlugins`, I think you will find the answer.

File: caddyhttp/httpserver/server.go (hunk 15 above)
Comment (Ngo-The-Trung) [inline caddyhttp/httpserver/server.go]: No. Could be anything. Should I use `interface{}` instead?

File: caddyhttp/maxrequestbody/maxrequestbody.go (hunk 12 above)
Comment (Ngo-The-Trung) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: Forgot to remove

File: caddyhttp/maxrequestbody/maxrequestbody.go (hunk 13 above)
Comment (Ngo-The-Trung) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: Well I think everything should have a pair, with the only exception being the first argument.

The 2 cases you mentioned would fail due to the ParseInt below, also I have no idea how to interpret them.

File: caddyhttp/maxrequestbody/maxrequestbody_test.go (hunk 10 above)
Comment (Ngo-The-Trung) [inline caddyhttp/maxrequestbody/maxrequestbody_test.go]: Not easy. 100 * KB is more readable than say 100 * validUnits["KB"] or 100 * validUnits[1], depending on whether I'm making it a map or a list.

Also I traverse the units in order in the parser code so if I were to rewrite validUnits to a map I would need to make an additional variable to store the order of the keys.

I would keep the constants.

File: caddyhttp/maxrequestbody/maxrequestbody.go (hunk 13 above)
Comment (tw4452852) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: Hmm, Is there any reason for this exception?

File: caddyhttp/httpserver/server.go (hunk 15 above)
Comment (tw4452852) [inline caddyhttp/httpserver/server.go]: Maybe `struct{}` better.

File: caddyhttp/httpserver/server.go (hunk 14 above)
Comment (Ngo-The-Trung) [inline caddyhttp/httpserver/server.go]: My intention is that the status code is advisory only, and is only applicable when the returned bool is `true`. It should be discarded if it is `false`.

File: caddyhttp/maxrequestbody/maxrequestbody.go (hunk 13 above)
Comment (Ngo-The-Trung) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: For convenience

`maxrequestbody 200MB` is understood as "limit everything (/) to 200MB"

`maxrequestbody 200MB /upload 500GB` is understood as "limit upload to 500GB, and limit everythign else to 200MB"

File: caddyhttp/maxrequestbody/maxrequestbody_test.go (hunk 10 above)
Comment (Ngo-The-Trung) [inline caddyhttp/maxrequestbody/maxrequestbody_test.go]: Edit: I meant to reply to the `validUnits` comment

File: caddyhttp/httpserver/server.go (hunk 14 above)
Comment (tw4452852) [inline caddyhttp/httpserver/server.go]: Make it clear in comment.

File: caddyhttp/maxrequestbody/maxrequestbody.go [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+	for i < len(args) {
+		path := args[i]
+		limitStr := args[i+1]
+		i = i + 2
+
+		if size := parseSize(limitStr); size < 1 { // also disallow size = 0
+			return pathLimit, errors.New("Parse failed")
+		} else {
+			pathLimit = addPathLimit(pathLimit, path, size)
+		}
```
Comment (elcore) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: You do not need to use else here

File: caddyhttp/maxrequestbody/maxrequestbody.go [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+	sizeStr = strings.ToUpper(sizeStr)
+
+	for _, unit := range validUnits {
+		if strings.HasSuffix(sizeStr, unit.symbol) {
+
+			if size, err := strconv.ParseInt(sizeStr[0:len(sizeStr)-len(unit.symbol)], 10, 64); err != nil {
+				return -1
+			} else {
+				return size * unit.multiplier
+			}
```
Comment (elcore) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: You do not need to use else here

File: caddyhttp/maxrequestbody/maxrequestbody.go [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+
+func (s *pathLimitSorter) Swap(i, j int) {
+	s.pathLimits[i], s.pathLimits[j] = s.pathLimits[j], s.pathLimits[i]
+}
+
+func (s *pathLimitSorter) Less(i, j int) bool {
+	return s.by(&s.pathLimits[i], &s.pathLimits[j])
+}
+
+func LengthDescending(p1, p2 *httpserver.PathLimit) bool {
```
Comment (elcore) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: We need some godoc here

File: caddyhttp/maxrequestbody/maxrequestbody_test.go [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+			},
+			hasError: false},
+		{
+			arguments: []string{"999", "/home", "12345MB"},
+			expected: []httpserver.PathLimit{
+				httpserver.PathLimit{Path: "/", Limit: 999},
+				httpserver.PathLimit{Path: "/home", Limit: 12345 * MB},
+			},
+			hasError: false,
+		},
```
Comment (elcore) [inline caddyhttp/maxrequestbody/maxrequestbody_test.go]: Please use gofmt

File: caddyhttp/httpserver/siteconfig.go (hunk 2 above)
Comment (elcore) [inline caddyhttp/httpserver/siteconfig.go]: Yes, please add some godoc.

File: caddyhttp/maxrequestbody/maxrequestbody_test.go (hunk 19 above)
Comment (Ngo-The-Trung) [inline caddyhttp/maxrequestbody/maxrequestbody_test.go]: My gofmt seems to be ok with this for some reason

Comment (Ngo-The-Trung) [inline caddyhttp/maxrequestbody/maxrequestbody_test.go]: Oh gofmt wasn't the reason. You guys are using simplify mode (-s).

I'll update this.

**Comment by Ngo-The-Trung** [comment]
Well at least the CI is green now

File: caddyhttp/maxrequestbody/maxrequestbody_test.go (hunk 19 above)
Comment (elcore) [inline caddyhttp/maxrequestbody/maxrequestbody_test.go]: Well, it´s part of gofmt 😉 -- Our code should be simple 😄

**Comment by elcore** [comment]
> Well at least the CI is green now

@Ngo-The-Trung mission accomplished 😄

**Comment by Ngo-The-Trung** [comment]
Pinging @mholt @tw4452852 for review 

Not sure how to test the limiting itself.

File: caddyhttp/maxrequestbody/maxrequestbody.go [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+
+// addPathLimit appends the path-to-request body limit mapping to pathLimit
+// Slashes are checked and added to path if necessary. Duplicates are ignored.
+func addPathLimit(pathLimit []httpserver.PathLimit, path string, limit int64) []httpserver.PathLimit {
+	// Enforces preceding slash
+	if path[0] != '/' {
+		path = "/" + path
+	}
+
+	// Avoid duplicates
```
Comment (tw4452852) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: I think it's better to overwrite the exist one.

File: caddyhttp/maxrequestbody/maxrequestbody_test.go [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+	for caseNum, c := range cases {
+		output, err := parseArguments(c.arguments)
+		if c.hasError && (err == nil) {
+			t.Errorf("Expecting error for case %v but none encountered", caseNum)
+		}
+		if !c.hasError && (err != nil) {
+			t.Errorf("Expecting no error for case %v but encountered %v", caseNum, err)
+		}
+
+		if reflect.DeepEqual(c.expected, len(output)) {
```
Comment (tw4452852) [inline caddyhttp/maxrequestbody/maxrequestbody_test.go]: Should be: `!reflect.DeepEqual(c.expected, output)`

File: caddyhttp/maxrequestbody/maxrequestbody_test.go [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+				{Path: "/static/images", Limit: 100},
+				{Path: "/static", Limit: 1},
+				{Path: "/index", Limit: 200}},
+		},
+	}
+
+	for caseNum, c := range cases {
+		output := append([]httpserver.PathLimit{}, c.arguments...)
+		SortPathLimits(output)
+		if reflect.DeepEqual(c.expected, len(output)) {
```
Comment (tw4452852) [inline caddyhttp/maxrequestbody/maxrequestbody_test.go]: Ditto.

File: caddyhttp/maxrequestbody/maxrequestbody_test.go (hunk 21 above)
Comment (Ngo-The-Trung) [inline caddyhttp/maxrequestbody/maxrequestbody_test.go]: Oh lol the 2 wrongs make it right

**Comment by Ngo-The-Trung** [comment]
Updated

**Comment by mholt** [comment]
Making progress! :) I'll look at it more when I'm back from my trip.

File: caddyhttp/httpserver/server.go [hunk 23]
```diff
@@ -406,3 +488,15 @@ func WriteTextResponse(w http.ResponseWriter, status int, body string) {
 	w.WriteHeader(status)
 	w.Write([]byte(body))
 }
+
+// IsCommonError checks if the given error is an error that could pop up
+// when handling request. Currently only checks for MaxBytesReader error.
+// Returns a boolean - whether the given error is a common error, and an int -
+// the advisory HTTP status code to return to the user (or -1 if the given error
+// is not a common error)
+func IsCommonError(err error) (bool, int) {
```
Comment (mholt) [inline caddyhttp/httpserver/server.go]: I still don't quite feel good about this function. It's only used by MaxBytesReader, I think error handling should not be here. It should probably remain in the handlers that use the MaxBytesReader.

Comment (Ngo-The-Trung) [inline caddyhttp/httpserver/server.go]: Ok

**Comment by Ngo-The-Trung** [comment]
Updated

File: caddyhttp/maxrequestbody/maxrequestbody.go (hunk 13 above)
Comment (mholt) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: I think this is a little confusing... A more Caddy-like way of doing pairs would be a "table" format. How about these valid forms instead:

```
maxrequestbody 200MB
maxrequestbody /upload 500GB
maxrequestbody {
    /ten 10MB
    /twenty 20MB
}
```

I think this way is much clearer.

File: caddyhttp/maxrequestbody/maxrequestbody.go [hunk 24]
```diff
@@ -0,0 +1,151 @@
...
+		path = "/" + path
+	}
+
+	// Use the last value if there are duplicates
+	for i, p := range pathLimit {
+		if p.Path == path {
+			pathLimit[i].Limit = limit
+			return pathLimit
+		}
+	}
```
Comment (mholt) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: Nice, glad you're thinking of stuff like this. :+1:

File: caddyhttp/proxy/proxy.go [hunk 25]
```diff
@@ -174,6 +174,10 @@ func (p Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error) {
 			return 0, nil
 		}
 
+		if _, ok := backendEerr.(httpserver.MaxBytesExceeded); ok {
+			return http.StatusRequestEntityTooLarge, backendErr
+		}
```
Comment (mholt) [inline caddyhttp/proxy/proxy.go]: _NICE._ This is what I was looking for. :tada:

**Comment by mholt** [comment]
Cool, I'm almost done with a final review -- looking pretty good so far (I might massage a few things after merging, no big deal). Just need a few more minutes. :) Perhaps meanwhile @tw4452852 and @elcore would be able to update their reviews or approve the changes.

File: caddyhttp/maxrequestbody/maxrequestbody_test.go [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+
+	"github.com/mholt/caddy/caddyhttp/httpserver"
+)
+
+const (
+	KB = 1024
+	MB = 1024 * 1024
+	GB = 1024 * 1024 * 1024
+)
+
```
Comment (tw4452852) [inline caddyhttp/maxrequestbody/maxrequestbody_test.go]: Add `TestSetupMaxRequestBody` ? (just test the all kinds of syntax formats)

File: caddyhttp/maxrequestbody/maxrequestbody_test.go [hunk 27]
```diff
@@ -0,0 +1,127 @@
...
+		arguments []pathLimitUnparsed
+		expected  []httpserver.PathLimit
+		hasError  bool
+	}{
+		// Parse errors
+		{arguments: []pathLimitUnparsed{{"/", "123.5"}}, expected: []httpserver.PathLimit{}, hasError: true},
+		{arguments: []pathLimitUnparsed{{"/", "200LB"}}, expected: []httpserver.PathLimit{}, hasError: true},
+		{arguments: []pathLimitUnparsed{{"/", "path:999MB"}}, expected: []httpserver.PathLimit{}, hasError: true},
+		{arguments: []pathLimitUnparsed{{"/", "1_234_567"}}, expected: []httpserver.PathLimit{}, hasError: true},
+
```
Comment (tw4452852) [inline caddyhttp/maxrequestbody/maxrequestbody_test.go]: Add test with zero size.

**Comment by Ngo-The-Trung** [comment]
@tw4452852 done

File: caddyhttp/maxrequestbody/maxrequestbody.go [hunk 28]
```diff
@@ -0,0 +1,178 @@
...
+
+	args := c.RemainingArgs()
+	argList := []pathLimitUnparsed{}
+
+	switch len(args) {
+	case 0:
+		// Format: { <path> <limit> ... }
+		for c.NextBlock() {
+			path := c.Val()
+			if !c.NextBlock() {
```
Comment (tw4452852) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: I think we should use `NextArg` here.

**Comment by mholt** [comment]
@tw4452852 When you are satisfied, approve and merge away!

File: caddyhttp/maxrequestbody/maxrequestbody.go (hunk 28 above)
Comment (Ngo-The-Trung) [inline caddyhttp/maxrequestbody/maxrequestbody.go]: Done

**Comment by Ngo-The-Trung** [comment]
@mholt moved to below "bind"

**Comment by tw4452852** [comment]
LGTM. Thanks @Ngo-The-Trung !

**Comment by mholt** [comment]
@Ngo-The-Trung Thank you for your work on this!

Would you like to be a collaborator? I'll send you the invite and you may accept it if you want. This means you'd be able to contribute to Caddy without having to fork the project (you could commit to branches) and you can help by reviewing PRs, etc. Really appreciate your efforts. :)
